### DISCUSSION TITLE
Dispose MorphTargets

### DISCUSSION THREAD

**Discussion by arpu** [donmccurdy/glTF-Transform#682] [Q&A] [answered]
Hey,

is there a way to list MorphTargets and dispose this?

**Comment by donmccurdy** [comment, marked as answer]
Morph targets are associated with mesh primitives, so you can remove them like this:

```js
import { prune } from '@gltf-transform/functions';

for (const mesh of document.getRoot().listMeshes()) {
  for (const prim of mesh.listPrimitives()) {
    prim.listTargets().forEach((target) => target.dispose());
  }
}

await document.transform(prune());
```

The `prune()` step will clean up any unused resources (e.g. vertex attributes) left behind by the removed morph targets.

You may additionally want to clean up any weights or names associated with the morph targets. Not all models will have these, but it doesn't hurt to check:

```js
document.getRoot().listNodes().forEach((node) => node.setWeights([]));

document.getRoot().listMeshes().forEach((mesh) => {
  mesh.setWeights([])
  const extras = mesh.getExtras();
  delete extras.targetNames;
  mesh.setExtras(extras);
});
```

**Comment by arpu** [comment]
this works fantastic! 

only problem i see now is the gltf Validation shows me: 

```MESH_INVALID_WEIGHTS_COUNT | The length of weights array (249) does not match the number of morph targets (0).```

**Reply by donmccurdy** [reply]
Good point! I've added a few more notes to the previous answer. 👍

**Reply by arpu** [reply]
Thanks!

**Comment by arpu** [comment]
to fix this i use 

```
for (const mesh of doc.getRoot().listMeshes()) {
                 if (mesh.getWeights().length > 0 )
                    mesh.setWeights([])
}

```
